### PR TITLE
SLVS-1194 - Handle added files, renamed files and deleted files

### PR DESCRIPTION
**Pull request by vnaskos-sonar**
Fixes [SLVS-1194](https://sonarsource.atlassian.net/browse/SLVS-1194)

[SLVS-1194]: https://sonarsource.atlassian.net/browse/SLVS-1194?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/Integration/LocalServices/FileTracker.cs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+                Encoding.UTF8.WebName,
+                f)).ToList();
+            
+            var removedFileUris = removedFiles.Select(f => new FileUri(f)).ToList();
+                
+            fileRpcSlCoreService.DidUpdateFileSystem(new DidUpdateFileSystemParams(
+                removedFileUris, clientFiles));
+        }
+
+        return Task.CompletedTask;
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Integration/LocalServices/FileTracker.cs:82]: Log debug if service not available?

File: src/Integration.UnitTests/MefServices/ProjectDocumentsEventsListenerTests.cs [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    {
+        var testSubject = CreateTestSubject();
+
+        var result = testSubject.OnQueryRenameFiles(null, 1, [], [], [], [], []);
+
+        result.Should().Be(VSConstants.S_OK);
+    }
+
+    [TestMethod]
+    public void OnQueryRenameDirectories_DoesNothing()
```
[the remark this answers is not in the document]
Comment (vnaskos-sonar) [reply src/Integration.UnitTests/MefServices/ProjectDocumentsEventsListenerTests.cs:135]: Unfortunately no!

It is triggered on directory rename, it does not contain the files of the directory and it does not trigger any file events either :/

[the remark this answers is not in the document]
Comment (georgii-borovinskikh-sonarsource) [reply src/Integration.UnitTests/MefServices/ProjectDocumentsEventsListenerTests.cs:135]: I've added a task for this

File: src/Integration.Vsix/Events/ProjectDocumentsEventsListener.cs [hunk 3]
```diff
@@ -24,28 +24,35 @@
...
-        trackProjectDocuments = (IVsTrackProjectDocuments2)serviceProvider.GetService(typeof(SVsTrackProjectDocuments));
     }
-    
+
     public void Initialize()
     {
-        trackProjectDocuments.AdviseTrackProjectDocumentsEvents(this, out trackDocumentEventsCookie);
+        threadHandling.RunOnUIThread(() =>
+        {
+            threadHandling.ThrowIfNotOnUIThread();
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Integration.Vsix/Events/ProjectDocumentsEventsListener.cs:52]: this is not needed right after RunOnUIThread

File: src/Integration.Vsix/Events/ProjectDocumentsEventsListener.cs [hunk 4]
```diff
@@ -24,28 +24,35 @@
...
         this.fileTracker = fileTracker;
         this.threadHandling = threadHandling;
-        trackProjectDocuments = (IVsTrackProjectDocuments2)serviceProvider.GetService(typeof(SVsTrackProjectDocuments));
     }
-    
+
     public void Initialize()
     {
-        trackProjectDocuments.AdviseTrackProjectDocumentsEvents(this, out trackDocumentEventsCookie);
+        threadHandling.RunOnUIThread(() =>
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Integration.Vsix/Events/ProjectDocumentsEventsListener.cs:50]: Use `IVsUIServiceOperation`

File: src/Integration.Vsix/Events/ProjectDocumentsEventsListener.cs [hunk 5]
```diff
@@ -134,18 +141,17 @@ public void Initialize()
...
     public void Dispose()
     {
-        Dispose(disposing: true);
-        GC.SuppressFinalize(this);
-    }
-    
-    protected virtual void Dispose(bool disposing)
-    {
-        if (disposing)
+        if (isDisposed)
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Integration.Vsix/Events/ProjectDocumentsEventsListener.cs:147]: What's your opinion on `if (disposed)` vs `if (isDisposed)`?

File: src/Integration.Vsix/Events/ProjectDocumentsEventsListener.cs [hunk 6]
```diff
@@ -24,28 +24,35 @@
...
-    
+
     public void Initialize()
     {
-        trackProjectDocuments.AdviseTrackProjectDocumentsEvents(this, out trackDocumentEventsCookie);
+        threadHandling.RunOnUIThread(() =>
+        {
+            threadHandling.ThrowIfNotOnUIThread();
+            trackProjectDocuments = serviceProvider.GetService(typeof(SVsTrackProjectDocuments)) as IVsTrackProjectDocuments2;
+            trackProjectDocuments?.AdviseTrackProjectDocumentsEvents(this, out trackDocumentEventsCookie);
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Integration.Vsix/Events/ProjectDocumentsEventsListener.cs:54]: You're not checking the result here. Use `Microsoft.VisualStudio.ErrorHandler` similar to ActiveSolutionTracker

File: src/Integration/LocalServices/FileTracker.cs [hunk 7]
```diff
@@ -58,26 +63,26 @@ public void RemoveFiles(params string[] removedFiles)
...
         {
-            var clientFiles = addedFiles.Select(f => new ClientFileDto(
-                new FileUri(f),
-                f.Substring(activeConfigScopeTracker.Current.RootPath.Length),
-                activeConfigScopeTracker.Current.Id,
-                null,
-                Encoding.UTF8.WebName,
-                f)).ToList();
-            
+            var rootPath = activeConfigScopeTracker.Current.RootPath;
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Integration/LocalServices/FileTracker.cs:74]: accessing `.Current` twice should be replaced by storing it in a local variable

File: src/Integration/LocalServices/FileTracker.cs [hunk 8]
```diff
@@ -58,26 +63,26 @@ public void RemoveFiles(params string[] removedFiles)
...
+            var clientFiles = addedFiles.Select(fp => clientFileDtoFactory.Create(fp, configScopeId, rootPath)).ToList();
             var removedFileUris = removedFiles.Select(f => new FileUri(f)).ToList();
-                
+
             fileRpcSlCoreService.DidUpdateFileSystem(new DidUpdateFileSystemParams(
                 removedFileUris, clientFiles));
         }
+        else
+        {
+            logger.WriteLine(Strings.FileTrackerGetRpcServiceError);
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Integration/LocalServices/FileTracker.cs:84]: Use `SLCoreStrings.ServiceProviderNotInitialized` if possible, maybe with some prefix to indicate where it happened. We've not been adding the prefix in other places, but maybe we should start to.

File: src/Integration/Resources/Strings.resx [hunk 9]
```diff
@@ -610,4 +610,7 @@ This is the general format to use when writing errors to output window or when t
   <data name="MuteWindow_WontFixTitle" xml:space="preserve">
     <value>Won't Fix</value>
   </data>
+  <data name="FileTrackerGetRpcServiceError" xml:space="preserve">
+    <value>FileTracker couldn't get transient service (IFileRpcSLCoreService)</value>
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Integration/Resources/Strings.resx:614]: string format + nameof would be better, but see the other comment

File: src/SLCore.UnitTests/Common/Helpers/ClientFileDtoFactoryTests.cs [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        var testSubject = new ClientFileDtoFactory();
+
+        var result = testSubject.Create(filePath, "CONFIG_SCOPE_ID", rootPath);
+        
+        ValidateDto(result, expectedUri, expectedFsPath, expectedIdeRelativePath);
+    }
+
+    private static void ValidateDto(ClientFileDto dto, string uri, string fsPath, string ideRelativePath)
+    {
+        dto.uri.ToString().Should().Be(uri);
```
Comment (georgii-borovinskikh-sonarsource) [inline src/SLCore.UnitTests/Common/Helpers/ClientFileDtoFactoryTests.cs:51]: You can use `FileUri.Equals` (via `Should().Be(...)`)instead of `.ToSting()`. Might remove the `string expectedUri` parameter as well.

File: src/SLCore.UnitTests/Common/Helpers/ClientFileDtoFactoryTests.cs [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+{
+    [TestMethod]
+    public void MefCtor_CheckIsSingleton()
+    {
+        MefTestHelpers.CheckIsSingletonMefComponent<ClientFileDtoFactory>();
+    }
+    
+    [DataTestMethod]
+    [DataRow("C:\\Code\\Project\\File1.js", "C:\\", "file:///C:/Code/Project/File1.js", "C:\\Code\\Project\\File1.js", "Code\\Project\\File1.js")]
+    [DataRow("C:\\привет\\project\\file1.js", "C:\\привет\\", "file:///C:/привет/project/file1.js", "C:\\привет\\project\\file1.js", "project\\file1.js")] // supports localized
```
Comment (georgii-borovinskikh-sonarsource) [inline src/SLCore.UnitTests/Common/Helpers/ClientFileDtoFactoryTests.cs:37]: I think copy-pasted tests with different names would be a fine alternative to comments, but it's up to you. Also, in NUnit you can add the description directly with the TestCaseAttribute. Can this be done with mstest?

File: src/SLCore.UnitTests/Common/Helpers/ClientFileDtoFactoryTests.cs [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+    public void Create_ConstructsValidDto(string filePath, string rootPath, string expectedUri, string expectedFsPath, string expectedIdeRelativePath)
+    {
+        var testSubject = new ClientFileDtoFactory();
+
+        var result = testSubject.Create(filePath, "CONFIG_SCOPE_ID", rootPath);
+        
+        ValidateDto(result, expectedUri, expectedFsPath, expectedIdeRelativePath);
+    }
+
+    private static void ValidateDto(ClientFileDto dto, string uri, string fsPath, string ideRelativePath)
```
Comment (georgii-borovinskikh-sonarsource) [inline src/SLCore.UnitTests/Common/Helpers/ClientFileDtoFactoryTests.cs:49]: these deserve `expected` prefix as well

File: src/SLCore.UnitTests/Common/Helpers/ClientFileDtoFactoryTests.cs [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    {
```
Comment (georgii-borovinskikh-sonarsource) [inline src/SLCore.UnitTests/Common/Helpers/ClientFileDtoFactoryTests.cs:50]: Might be possible to replace this whole thing with `.Should().BeEquivalentTo(new ClientFileDto(...))`

File: src/SLCore.UnitTests/Common/Helpers/ClientFileDtoFactoryTests.cs [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+        MefTestHelpers.CheckIsSingletonMefComponent<ClientFileDtoFactory>();
+    }
+    
+    [DataTestMethod]
+    [DataRow("C:\\Code\\Project\\File1.js", "C:\\", "file:///C:/Code/Project/File1.js", "C:\\Code\\Project\\File1.js", "Code\\Project\\File1.js")]
+    [DataRow("C:\\привет\\project\\file1.js", "C:\\привет\\", "file:///C:/привет/project/file1.js", "C:\\привет\\project\\file1.js", "project\\file1.js")] // supports localized
+    [DataRow("\\\\servername\\work\\project\\file1.js", "\\\\servername\\work\\", "file://servername/work/project/file1.js", "\\\\servername\\work\\project\\file1.js", "project\\file1.js")] // supports UNC
+    [DataRow("C:\\Code\\My Project\\My Favorite File2.js", "C:\\", "file:///C:/Code/My%20Project/My%20Favorite%20File2.js", "C:\\Code\\My Project\\My Favorite File2.js", "Code\\My Project\\My Favorite File2.js")] // supports whitespaces
+    public void Create_ConstructsValidDto(string filePath, string rootPath, string expectedUri, string expectedFsPath, string expectedIdeRelativePath)
```
Comment (georgii-borovinskikh-sonarsource) [inline src/SLCore.UnitTests/Common/Helpers/ClientFileDtoFactoryTests.cs:40]: Not enough relative path tests. That's one of the main responsibilities of this class.

File: src/SLCore/Common/Helpers/ClientFileDtoFactory.cs [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    public ClientFileDto Create(string fullPath, string configScopeId, string rootPath)
+    {
+        var ideRelativePath = GetRelativePath(rootPath, fullPath);
+        var uri = new FileUri(fullPath);
+        return new ClientFileDto(uri, ideRelativePath, configScopeId, null, Encoding.UTF8.WebName, fullPath);
+    }
+
+    private static string GetRelativePath(string root, string fullPath)
+    {
+        return fullPath.Substring(root.Length);
```
Comment (georgii-borovinskikh-sonarsource) [inline src/SLCore/Common/Helpers/ClientFileDtoFactory.cs:41]: Too few tests for this

File: src/Integration.Vsix.UnitTests/Events/ProjectDocumentsEventsListenerTests.cs [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+
+        var result = testSubject.OnAfterSccStatusChanged(1, 1, [], [], [], []);
+
+        result.Should().Be(VSConstants.S_OK);
+    }
+
+    [TestMethod]
+    public void Dispose_UnadviseTrackProjectDocumentsEvents()
+    {
+        var trackProjectDocuments2 = Substitute.For<IVsTrackProjectDocuments2>();
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Integration.Vsix.UnitTests/Events/ProjectDocumentsEventsListenerTests.cs:199]: This is interesting. You're using a mock, but then the test failed with a real exception

```
Test method SonarLint.VisualStudio.Integration.UnitTests.Events.ProjectDocumentsEventsListenerTests.Dispose_UnadviseTrackProjectDocumentsEvents threw exception:
System.Runtime.InteropServices.COMException: Dispose must be called on the UI thread.. If you are using UI objects in test consider using [UITestMethod] attribute instead of [TestMethod] to execute test in UI thread.


```

File: src/Integration.Vsix/Events/ProjectDocumentsEventsListener.cs [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+    }
+
+    public void Dispose()
+    {
+        if (isDisposed)
+        {
+            return;
+        }
+     
+        ThreadHelper.ThrowIfNotOnUIThread();
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Integration.Vsix/Events/ProjectDocumentsEventsListener.cs:152]: Dispose is not guaranteed to be called on the UI thread

File: src/Integration.Vsix.UnitTests/Events/ProjectDocumentsEventsListenerTests.cs [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+
+        trackProjectDocuments2.Received().UnadviseTrackProjectDocumentsEvents(Arg.Any<uint>());
+    }
+
+    private static ProjectDocumentsEventsListener CreateTestSubject(IFileTracker fileTracker = null,
+        IVsTrackProjectDocuments2 trackProjectDocuments2 = null, IThreadHandling threadHandling = null)
+    {
+        trackProjectDocuments2 ??= Substitute.For<IVsTrackProjectDocuments2>();
+        var serviceProvider = Substitute.For<IServiceProvider>();
+        serviceProvider.GetService(typeof(SVsTrackProjectDocuments)).Returns(trackProjectDocuments2);
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Integration.Vsix.UnitTests/Events/ProjectDocumentsEventsListenerTests.cs:212]: Setup should be avoided in CreateTestSubject

File: src/Integration.Vsix/Events/ProjectDocumentsEventsListener.cs (hunk 5 above)
Comment (vnaskos-sonar) [reply src/Integration.Vsix/Events/ProjectDocumentsEventsListener.cs:147]: I don't have any strong preferences, as long as we are consistent with the rest of the codebase, it should be fine. Here I think is also better the isDisposed because disposed is used as a param of Dispose(bool dispose).

isDisposed is used in ActiveSolutionTracker

File: src/Integration.Vsix/SonarLintTagger/TextBufferIssueTracker.cs [hunk 19]
```diff
@@ -71,9 +72,9 @@ internal sealed class TextBufferIssueTracker : IIssueTracker, ITagger<IErrorTag>
             this.logger = logger;
 
             this.document = document;
-            this.LastAnalysisFilePath = document.FilePath;
-
-            this.Factory = new IssuesSnapshotFactory(LastAnalysisFilePath);
+            LastAnalysisFilePath = document.FilePath;
+            fileTracker.AddFiles(LastAnalysisFilePath);
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Integration.Vsix/SonarLintTagger/TextBufferIssueTracker.cs:76]: Can we also update the file encoding here? Currently on ListFiles we always use the UTF8 default. It is passed to the SLCore analyzer for legacy reasons, but it can't be passed to the `analyzeFilesAndTrack` method of sloop. We might update the encoding here instead and later, when we move all languages to sloop, remove it from the analyzer interface

Comment (georgii-borovinskikh-sonarsource) [reply src/Integration.Vsix/SonarLintTagger/TextBufferIssueTracker.cs:76]: If it would take a considerable amount of time to do this, then just create a ticket

Comment (vnaskos-sonar) [reply src/Integration.Vsix/SonarLintTagger/TextBufferIssueTracker.cs:76]: https://sonarsource.atlassian.net/browse/SLVS-1219

File: src/Integration/LocalServices/FileTracker.cs [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+            var configScope = activeConfigScopeTracker.Current;
+            var clientFiles = addedFiles.Select(fp => clientFileDtoFactory.Create(fp, configScope.Id, configScope.RootPath)).ToList();
+            var removedFileUris = removedFiles.Select(f => new FileUri(f)).ToList();
+
+            fileRpcSlCoreService.DidUpdateFileSystem(new DidUpdateFileSystemParams(
+                removedFileUris, clientFiles));
+        }
+        else
+        {
+            logger.WriteLine("[FileTracker] {0}", SLCoreStrings.ServiceProviderNotInitialized);
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Integration/LocalServices/FileTracker.cs:83]: Use `nameof`. Also, I'm wondering if it deserves a method overload (to automate future contextual logs with prefixes). Add a jira ticket

File: src/vs-threading.MainThreadSwitchingMethods.txt [hunk 21]
```diff
@@ -3,5 +3,4 @@
 
 [SonarLint.VisualStudio.Core.IThreadHandling]::SwitchToMainThreadAsync
 [SonarLint.VisualStudio.Core.IThreadHandling]::RunOnUIThread
-[SonarLint.VisualStudio.Core.IThreadHandling]::RunOnUIThreadAsync
-
+[SonarLint.VisualStudio.Core.IThreadHandling]::RunOnUIThreadAsync
```
Comment (georgii-borovinskikh-sonarsource) [inline src/vs-threading.MainThreadSwitchingMethods.txt:6]: Can probably just revert this change to remove it from the diff

File: src/SLCore/Common/Helpers/ClientFileDtoFactory.cs (hunk 15 above)
Comment (georgii-borovinskikh-sonarsource) [reply src/SLCore/Common/Helpers/ClientFileDtoFactory.cs:41]: This still stands

**Review by georgii-borovinskikh-sonarsource — APPROVED** [review]
A few small changes still left. You can merge this after you fix those

File: src/Integration/LocalServices/FileTracker.cs [hunk 22]
```diff
@@ -0,0 +1,88 @@
...
+        }
+
+        return Task.CompletedTask;
```
[the remark this answers is not in the document]
Comment (vnaskos-sonar) [reply src/Integration/LocalServices/FileTracker.cs:86]: Although that would have been nice, adding async complains that we don't have an await and none of the calls inside the method are async.
